nomos-services/api/src/http/da_shares.rs

```diff
@@ -0,0 +1,142 @@
+use std::{collections::HashSet, hash::Hash, io::ErrorKind, sync::Arc};
+
+use bytes::Bytes;
+use futures::{stream, Stream, StreamExt};
+use nomos_core::da::blob::{LightShare, Share};
+use nomos_da_storage::rocksdb::{
+    create_share_idx, key_bytes, DA_BLOB_SHARES_INDEX_PREFIX, DA_SHARE_PREFIX,
+};
+use nomos_storage::{
+    backends::{rocksdb::RocksBackend, StorageSerde},
+    StorageMsg, StorageService,
+};
+use overwatch::{overwatch::handle::OverwatchHandle, services::relay::OutboundRelay};
+use serde::{de::DeserializeOwned, Serialize};
+use tokio::sync::oneshot;
+
+#[expect(clippy::implicit_hasher, reason = "Don't need custom hasher")]
+pub async fn get_shares<StorageOp, DaShare>(
+    handle: &OverwatchHandle,
+    blob_id: DaShare::BlobId,
+    requested_shares: HashSet<DaShare::ShareIndex>,
+    filter_shares: HashSet<DaShare::ShareIndex>,
+    return_available: bool,
+) -> Result<
+    impl Stream<Item = Result<Bytes, crate::http::DynError>> + Send + Sync + 'static,
+    crate::http::DynError,
+>
+where
+    StorageOp: StorageSerde + Send + Sync + 'static,
+    DaShare: Share + 'static,
+    DaShare::BlobId: AsRef<[u8]> + Clone + DeserializeOwned + Send + Sync + 'static,
+    DaShare::ShareIndex: AsRef<[u8]> + DeserializeOwned + Hash + Eq + Send + Sync + 'static,
+    DaShare::LightShare: LightShare<ShareIndex = DaShare::ShareIndex>
+        + Serialize
+        + DeserializeOwned
+        + Send
+        + Sync
+        + 'static,
+    StorageOp::Error: std::error::Error + Send + Sync + 'static,
+{
+    let storage_relay = handle
+        .relay::<StorageService<RocksBackend<StorageOp>>>()
+        .connect()
+        .await?;
+
+    let blob_shares =
+        load_blob_shares_index::<StorageOp, DaShare>(&storage_relay, &blob_id).await?;
+
+    let filtered_shares: Vec<_> = blob_shares
+        .into_iter()
+        .filter(|idx| {
+            // If requested_shares contains the index, then ignore the filter_shares
+            requested_shares.contains(idx) || (return_available && !filter_shares.contains(idx))
+        })
+        .collect();
```

Comment:
```suggestion
    let filtered_shares = blob_shares.into_iter().filter(move |idx| {
        // If requested_shares contains the index, then ignore the filter_shares
        requested_shares.contains(idx) || (return_available && !filter_shares.contains(idx))
    });
```

Since we're streaming the iterator later on, I think `collect` is not required here as `Filter` is also an `Iterator`

Reply:
Thanks! Removed unnecessary `collect`